Apps/Src/ReadCarCAN.c

```diff
@@ -52,114 +76,145 @@ static inline void chargingEnable(void) {
 {
     OS_ERR err;
 
     //data struct for CAN message
     CANDATA_t dataBuf;
-    CPU_TS ts;
 
     OSMutexCreate(&arrayRestartMutex, "array restart mutex", &err);
     assertOSError(OS_READ_CAN_LOC,err);
-    OSMutexCreate(&msg_rcv_mutex, "message received mutex", &err);
-    assertOSError(OS_READ_CAN_LOC,err);
 
-    //Create+start the Can watchdog thread
-    OSTaskCreate(
-        &canWatchTCB,
-        "CAN Watchdog",
-        &CANWatchdog_Handler,
-        NULL,
-        TASK_READ_CAR_CAN_PRIO,
-        canWatchSTK,
-        DEFAULT_STACK_SIZE/10,
-        DEFAULT_STACK_SIZE,
+
+    // Create the CAN Watchdog (periodic) timer, which disconnects the array and disables regenerative braking
+    // if we do not get a CAN message with the ID Charge_Enable within the desired interval.
+    OSTmrCreate(
+        &canWatchTimer,
+        "CAN Watch Timer",
         0,
+        CAN_WATCH_TMR_DLY_TMR_TS,
+        OS_OPT_TMR_PERIODIC,
+        canWatchTimerCallback,
+        NULL,
+        &err
+    );
+    assertOSError(OS_READ_CAN_LOC, err);
+
+    OSTmrCreate(
+        &prechargeDlyTimer,
+        "Precharge Delay Timer",
+        PRECHARGE_DLY_TMR_TS,
         0,
+        OS_OPT_TMR_ONE_SHOT,
+        arrayRestart,
         NULL,
-        OS_OPT_TASK_STK_CLR,
         &err
     );
-    assertOSError(OS_READ_CAN_LOC,err);
+    assertOSError(OS_READ_CAN_LOC, err);
+
+    //Start CAN Watchdog timer
+    OSTmrStart(&canWatchTimer, &err);
+    assertOSError(OS_READ_CAN_LOC, err);
+    
 
     while (1)
     {
+        
         //Get any message that BPS Sent us
         ErrorStatus status = CANbus_Read(&dataBuf,CAN_BLOCKING,CARCAN);  
         if(status != SUCCESS) {
             continue;
         }
 
         switch(dataBuf.ID){ //we got a message
-            case CHARGE_ENABLE: {
-                OSMutexPend(&msg_rcv_mutex,
-                    0,
-                    OS_OPT_PEND_BLOCKING,
-                    &ts,
-                    &err);
-                assertOSError(OS_READ_CAN_LOC,err);
+            case CHARGE_ENABLE: { 
 
-                msg_recieved = true; //signal success recieved
+                // Restart CAN Watchdog timer
+                OSTmrStart(&canWatchTimer, &err);
+                assertOSError(OS_READ_CAN_LOC, err);
 
-                OSMutexPost(&msg_rcv_mutex,
-                            OS_OPT_NONE,
-                            &err);
-                assertOSError(OS_READ_CAN_LOC,err);
 
                 if((&dataBuf.data)[0] == 0){ // If the buffer doesn't contain 1 for enable, turn off RegenEnable and turn array off
                     chargingDisable();
-                } else {
-                    //We got a message of enable with a nonzero value, turn on Regen, If we are already in precharge / array is on, do nothing. 
-                    //If not initiate precharge and restart sequence. 
-                    chargingEnable();
+                    updateSaturation(-1); // Update saturation and buffer
```

Comment:
`chargeMsgSaturation` isn't initialized anywhere, so it'll start off as zero. Is this intentional, or should it start off near the threshold?

Reply:
It wasn't intentional, but I think initializing it to zero works since the array used to calculate it will start off as all zeros as well, so chargeMsgSaturation will end up becoming zero as soon as we update it. 

I went back and explicitly initialized it to zero so that it's clearer, though